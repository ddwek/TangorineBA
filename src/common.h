/*
 * Copyright 2025 Daniel Dwek
 *
 * This file is part of TangorineBA.
 *
 *  TangorineBA is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TangorineBA is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TangorineBA.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#ifndef _COMMON_H_
#define _COMMON_H_	1
#include <gtk/gtk.h>

// Bitmask of flags for each cell
typedef struct bm_flags_st {
	unsigned int imm : 1;		// Immutable
	unsigned int top : 1;		// Top constraint
	unsigned int top_equal : 1;	// Top must be equal (1) or diff (0) to prev row cell
	unsigned int right : 1;
	unsigned int right_equal : 1;
	unsigned int bottom : 1;
	unsigned int bottom_equal : 1;
	unsigned int left : 1;
	unsigned int left_equal : 1;	// ...
	unsigned int padding : 23;	// Reserved amount of bits up to complete 32-bit int
	int claim_for_hor_hatching;	// Horizontal hatching
	int claim_for_ver_hatching;	// Vertical hatching
} bm_flags_t;

// Constraints' struct
typedef struct cons_st {
	int ncell;
	bm_flags_t flags;
	bm_flags_t& operator= (bm_flags_t&);	// Overload operator = for easy copy of flags
} cons_t;

// Input parsed info only meaningfull for testing (i.e., `make check`)
typedef struct parsed_st {
	int seed;		// Random seed
	int imm[6];		// Just 6 cells will be immutable
	cons_t cons[8];		// Just 8 constraints per board, but one cell can have more than one constraint
} in_parsed_t;

typedef enum { SHAPE_SUN = 0, SHAPE_MOON, SHAPE_EMPTY } shape_t;

typedef struct pending_events_st {
	int ncell;
	shape_t shape;
	bm_flags_t flags;
} pending_events_t;

typedef struct history_st {
	pending_events_t pe;
	int uid;
} history_t;

extern GtkWidget *undo_btn, *redo_btn;
#endif
