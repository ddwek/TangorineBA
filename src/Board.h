/*
 * Copyright 2025 Daniel Dwek
 *
 * This file is part of TangorineBA.
 *
 *  TangorineBA is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TangorineBA is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TangorineBA.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#ifndef _BOARD_H_
#define _BOARD_H_	1
#include <set>
#include <string>
#include <gtk/gtk.h>
#include "common.h"

typedef struct shape_info_st {
	int ncell;
	shape_t shape;
	bm_flags_t flags;
} shape_info_t;

class Board {
public:
	typedef enum { ROW = 0, COL } line_type_check;
	typedef enum { EQUAL = 0, DIFF } type_cons_t;
	typedef struct constr_st {
		int cell_0;
		int cell_1;
		type_cons_t constr;
	} constr_t;

	Board ();
	Board (bool testing, std::string test_filename);
	Board (Board&) = delete;
	Board (Board&&) = delete;
	Board& operator= (Board&) = delete;
	~Board ();

	void new_game ();
	void set_seed (int seed);
	void draw_shape (int nrow, int ncol, shape_t shape);
	void draw_cells (cairo_t *cr);
	bool can_draw_hatching (int ncell);
	void set_hatching (int ncell, bool hor);
	void clear_hatching (int ncell, bool hor);
	void draw_hatching (int ncell);
	void draw_hatching_on_immutable ();
	bool get_game_over ();
	void set_game_over (bool game_over);
	void show_congrats ();
	void change_shape (int nrow, int ncol);
	void change_row (int nrow, shape_t shape);
	void change_col (int ncol, shape_t shape);
	shape_info_t get_shape_status (int ncell) const;
	shape_info_t get_user_guess (int ncell) const;
	shape_info_t get_standard_solution (int ncell) const;
	void set_shape_status (int ncell, shape_t sh);
	void set_user_guess (int ncell, shape_t sh, bm_flags_t flags);
	int get_num_hsuns (int row, bool std);
	int get_num_vsuns (int col, bool std);
	int get_num_hmoons (int row, bool std);
	int get_num_vmoons (int col, bool std);
	int get_third_adjacent (int row, int col, line_type_check check, bool std);
	int is_valid (int *row, int *col, int *nsuns, int *nmoons, bool std);

	void validate_row_three_adjs (int nrow);
	void validate_row_diff_num_of_shapes (int nrow);
	void validate_row_constraints (int nrow);
	void validate_row (int nrow);
	void validate_col_three_adjs (int ncol);
	void validate_col_diff_num_of_shapes (int ncol);
	void validate_col_constraints (int ncol);
	void validate_col (int ncol);

	void prepare ();
	void print (bool is_testing, bool display_values, int n_step);
	std::string get_debug (int n_step, int row, int col) const;
	bool is_immutable (int n) const;
	void set_immutable_cells (int *imm);
	void set_immutable_cells ();
	void draw_immutable_cells ();
	void decode_flags (shape_info_t& ref);
	void set_constraints (cons_t *cons);
	void set_constraints ();
	void draw_constraints ();
	bool is_configured () const;
	bool is_testing () const;

private:
	cairo_t *cr;
	shape_info_t standard_solution[6][6];
	shape_info_t user_guess[6][6];
	bool configured;
	bool testing;
	std::string test_filename;
	std::string exp_filename;
	int seed;
	std::string debug[0x40][6][6];
	bool game_over;
};

extern class Board board;
#endif
