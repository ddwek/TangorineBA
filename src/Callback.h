/*
 * Copyright 2025 Daniel Dwek
 *
 * This file is part of TangorineBA.
 *
 *  TangorineBA is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TangorineBA is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TangorineBA.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#ifndef _CALLBACK_H_
#define _CALLBACK_H_	1
#include <gtk/gtk.h>
#include "common.h"

typedef struct region_st {
	int x0;
	int y0;
	int x1;
	int y1;
} cell_region_t;

class CallbackData {
public:
	CallbackData ();
	CallbackData (CallbackData&) = delete;
	CallbackData (CallbackData&&) = delete;
	CallbackData& operator= (CallbackData&) = delete;
	~CallbackData () = default;

	GtkApplication *get_app () const;
	GtkWindow *get_window () const;
	cairo_t *get_cr () const;
	cairo_t *get_timer_cr () const;
	cell_region_t *get_region (int n);
	int get_seconds () const;
	int get_minutes () const;

	void set_app (GtkApplication *app);
	void set_window (GtkWindow *window);
	void set_cr (cairo_t *cr);
	void set_timer_cr (cairo_t *timer_cr);
	void set_region (int nregion, int x0, int y0, int x1, int y1);
	void set_seconds (int seconds);
	void set_minutes (int minutes);

private:
	GtkApplication *app;
	GtkWindow *window;
	cairo_t *cr;
	cairo_t *timer_cr;
	cell_region_t region[36];
	int seconds;
	int minutes;
};

int on_tick_cb (GtkWidget *widget, GdkFrameClock *frame_clock, gpointer user_data);
int draw_timer_cb (GtkWidget *widget, cairo_t *timer_cr, gpointer data);
void clear_game_cb (GtkButton *btn);
void new_game_cb (GtkButton *btn);
int configure_cb (GtkWidget *widget, GdkEventConfigure *event, void *data);
int draw_cb (GtkWidget *widget, cairo_t *cr, gpointer data);
bool button_press_cb (GtkWidget *widget, GdkEventButton *button, void *data);
#endif
