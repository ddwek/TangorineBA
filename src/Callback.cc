/*
 * Copyright 2025 Daniel Dwek
 *
 * This file is part of TangorineBA.
 *
 *  TangorineBA is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TangorineBA is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TangorineBA.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include <string>
#include <list>
#include <set>
#include <gtk/gtk.h>
#include <gdk/gdk.h>
#include "common.h"
#include "Board.h"
#include "../test/Test.h"

extern class Test test;

bool are_there_pending_events = false;
typedef struct pending_events_st {
	int ncell;
	shape_t shape;
	bm_flags_t flags;
} pending_events_t;
std::list<pending_events_t> redraw_cells;

typedef struct region_st {
	int x0;
	int y0;
	int x1;
	int y1;
} cell_region_t;

class CallbackData {
public:
	CallbackData () = default;
	CallbackData (CallbackData&) = delete;
	CallbackData (CallbackData&&) = delete;
	CallbackData& operator= (CallbackData&) = delete;
	~CallbackData () = default;

	GtkApplication *get_app () const;
	GtkWindow *get_window () const;
	cairo_t *get_cr () const;
	cell_region_t *get_region (int n);

	void set_app (GtkApplication *app);
	void set_window (GtkWindow *window);
	void set_cr (cairo_t *cr);
	void set_region (int nregion, int x0, int y0, int x1, int y1);

private:
	GtkApplication *app;
	GtkWindow *window;
	cairo_t *cr;
	cell_region_t region[36];
} cbdata;

GtkApplication *CallbackData::get_app () const
{
	return app;
}

GtkWindow *CallbackData::get_window () const
{
	return window;
}

cairo_t *CallbackData::get_cr () const
{
	return cr;
}

cell_region_t *CallbackData::get_region (int n)
{
	if (n < 36)
		return &region[n];

	return nullptr;
}

void CallbackData::set_app (GtkApplication *app)
{
	this->app = app;
}

void CallbackData::set_window (GtkWindow *window)
{
	this->window = window;
}

void CallbackData::set_cr (cairo_t *cr)
{
	this->cr = cr;
}

void CallbackData::set_region (int n, int x0, int y0, int x1, int y1)
{
	region[n].x0 = x0;
	region[n].y0 = y0;
	region[n].x1 = x1;
	region[n].y1 = y1;
}

int configure_cb (GtkWidget *widget, GdkEventConfigure *event, void *data)
{
	for (int i = 0; i < 6; i++)
		for (int j = 0; j < 6; j++)
			cbdata.set_region (i * 6 + j, j * 80, i * 80, (j + 1) * 80, (i + 1) * 80);
	return 0;
}

int draw_cb (GtkWidget *widget, cairo_t *cr, void *user_data)
{
	struct _GdkRGBA bgcolor = { 0.2, 0.3, 0.4, 1.0 };
	std::list<pending_events_t>::iterator iter;

	cairo_save (cr);
	cbdata.set_cr (cr);
	gdk_cairo_set_source_rgba (cr, &bgcolor);
	cairo_paint (cr);
	board.draw_cells (cr);
	if (are_there_pending_events) {
		for (iter = redraw_cells.begin (); iter != redraw_cells.end (); iter++) {
			board.draw_immutable_cells ();
			board.draw_shape (iter->ncell / 6, iter->ncell % 6, iter->shape);
			if (board.can_draw_hatching (iter->ncell)) {
				board.draw_hatching (iter->ncell);
				board.draw_hatching_on_immutable ();
			}
		}

		if (board.get_game_over ())
			board.show_congrats ();
		are_there_pending_events = false;
	}
	board.draw_constraints ();
	cairo_restore (cr);

	return 0;
}

bool button_press_cb (GtkWidget *widget, GdkEventButton *event, void *user_data)
{
	static std::set<int> s;
	shape_t new_guess = SHAPE_EMPTY;
	pending_events_t pending_event;
	std::set<int>::iterator iter;

	for (int i = 0; i < 36; i++) {
		if (event->x > cbdata.get_region (i)->x0 && event->x < cbdata.get_region (i)->x1 &&
		    event->y > cbdata.get_region (i)->y0 && event->y < cbdata.get_region (i)->y1) {
			switch (board.get_user_guess (i).shape) {
			case SHAPE_SUN:
				new_guess = SHAPE_MOON;
				break;
			case SHAPE_MOON:
				new_guess = SHAPE_EMPTY;
				break;
			case SHAPE_EMPTY:
				new_guess = SHAPE_SUN;
				break;
			};

			if (!board.is_immutable (i)) {
				are_there_pending_events = true;
				pending_event.ncell = i;
				pending_event.shape = new_guess;
				pending_event.flags = board.get_standard_solution (i).flags;
				board.set_user_guess (i, new_guess, pending_event.flags);
				redraw_cells.push_back (pending_event);
			} else {
				are_there_pending_events = false;
			}

			board.validate_row (i / 6);
			board.validate_col (i % 6);
			if (are_there_pending_events) {
				if (board.get_user_guess (i).shape != SHAPE_EMPTY)
					s.insert (i);
empty_shape_erased:
				for (iter = s.begin (); iter != s.end (); iter++) {
					if (board.get_user_guess (*iter).shape == SHAPE_EMPTY) {
						// Caught *iter with an empty shape! Erasing...
						s.erase (iter);
						goto empty_shape_erased;
					}
				}
				if (s.size () == 30)
					board.set_game_over (true);
				gtk_widget_queue_draw (widget);
			}
		}
	}

	return true;
}
