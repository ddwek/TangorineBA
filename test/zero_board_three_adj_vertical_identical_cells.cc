/*
 * Copyright 2025 Daniel Dwek
 *
 * This file is part of TangorineBA.
 *
 *  TangorineBA is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  TangorineBA is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with TangorineBA.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include <list>
#include <gtk/gtk.h>
#include "../src/common.h"
#include "../src/Board.h"
#include "../src/Callback.h"
#include "Test.h"

extern bool are_there_pending_events;
extern std::list<pending_events_t> redraw_cells;
extern pending_events_t pending_event;
extern class CallbackData cbdata;
extern class Test test;
Board board (true, "./zero_board_three_adj_vertical_identical_cells");

int on_tick_cb (GtkWidget *widget, GdkFrameClock *frame_clock, gpointer data)
{
	static int i = 3, cnt = 0;
	static shape_t shape_0 = SHAPE_MOON, shape_1 = SHAPE_MOON, shape_2 = SHAPE_MOON;
	bool finished = false;
	int ret = -1;

	if (!board.is_configured ())
		if (!test.is_input_already_parsed () || !test.is_expected_already_parsed ())
			return -1;

	if (!(gdk_frame_clock_get_frame_counter (frame_clock) & 0xf)) {
		if (cnt / 3 == 0) {
			shape_0 = SHAPE_SUN;
			shape_1 = SHAPE_SUN;
			shape_2 = SHAPE_SUN;
		} else if (cnt / 3 == 1) {
			shape_0 = SHAPE_MOON;
			shape_1 = SHAPE_MOON;
			shape_2 = SHAPE_MOON;
		} else if (cnt / 3 == 2) {
			shape_0 = SHAPE_EMPTY;
			shape_1 = SHAPE_EMPTY;
			shape_2 = SHAPE_EMPTY;
		}

		shape_info_t first = board.get_user_guess (i * 6 + 6 * 0);
		pending_event.ncell = i * 6 + 6 * 0;
		pending_event.shape = shape_0;
		pending_event.flags = first.flags;
		board.set_user_guess (i * 6 + 6 * 0, pending_event.shape, pending_event.flags);
		redraw_cells.push_back (pending_event);
		cnt++;

		shape_info_t second = board.get_user_guess (i * 6 + 6 * 1);
		pending_event.ncell = i * 6 + 6 * 1;
		pending_event.shape = shape_1;
		pending_event.flags = second.flags;
		board.set_user_guess (i * 6 + 6 * 1, pending_event.shape, pending_event.flags);
		redraw_cells.push_back (pending_event);
		cnt++;

		shape_info_t third = board.get_user_guess (i * 6 + 6 * 2);
		pending_event.ncell = i * 6 + 6 * 2;
		pending_event.shape = shape_2;
		pending_event.flags = third.flags;
		board.set_user_guess (i * 6 + 6 * 2, pending_event.shape, pending_event.flags);
		redraw_cells.push_back (pending_event);
		cnt++;

		are_there_pending_events = true;
		if (cnt == 12)
			finished = true;

		for (int t = 0; t < 6; t++)
			board.validate_row (t);
		for (int t = 0; t < 6; t++)
			board.validate_col (t);
		gtk_widget_queue_draw (widget);

		if ((cnt / 3 - 1) < test.get_total_steps_to_parse ()) {
			board.print (true, true, cnt / 3 - 1);
			ret = test.pass (cnt / 3 - 1);
		}
	} else {
		gdk_frame_clock_get_frame_counter (frame_clock);
	}

	if (finished) {
		if (ret)
			exit (EXIT_SUCCESS);
		else
			exit (EXIT_FAILURE);
	}

	return ret;
}

int configure_testing_cb (GtkWidget *widget, GdkEventConfigure *event, void *data)
{
	for (int i = 0; i < 6; i++)
		for (int j = 0; j < 6; j++)
			cbdata.set_region (i * 6 + j, j * 80, i * 80, (j + 1) * 80, (i + 1) * 80);

	in_parsed_t *p = test.get_input_parsed ();
	board.set_seed (p->seed);
	board.set_immutable_cells (p->imm);
	gtk_widget_add_tick_callback (GTK_WIDGET (widget), on_tick_cb, nullptr, nullptr);
	return 0;
}

void activate (GtkApplication *app, void *data)
{
	GtkWidget *win;
	GtkWidget *da;

	win = gtk_application_window_new (app);
	da = gtk_drawing_area_new ();
	gtk_widget_set_size_request (da, 480, 480);
	gtk_container_add (GTK_CONTAINER (win), da);

	g_signal_connect (win, "destroy", G_CALLBACK (gtk_widget_destroy), win);
	g_signal_connect (da, "configure-event", G_CALLBACK (configure_testing_cb), nullptr);
	g_signal_connect (da, "draw", G_CALLBACK (draw_cb), nullptr);
	g_signal_connect (da, "button-press-event", G_CALLBACK (button_press_cb), nullptr);
	gtk_widget_set_events (da, gtk_widget_get_events (da) | GDK_BUTTON_PRESS_MASK);

	gtk_widget_show_all (win);
}

int main (int argc, char **argv)
{
	GtkApplication *app;
	int status;

	app = gtk_application_new ("org.gtk.TangorineBA", G_APPLICATION_DEFAULT_FLAGS);
	g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
	status = g_application_run (G_APPLICATION (app), argc, argv);
	g_object_unref (app);

	return status;
}
