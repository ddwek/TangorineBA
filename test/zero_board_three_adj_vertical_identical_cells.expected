n_steps = 3
steps: 1, 2, 3,
_[00m _[00m _[00m _[00m [01;44;37mD[00m _[00m 
[00;43;30m*[00m _[00m _[00m _[00m _[00m _[00m 
_[00m [00;43;30m*[00m _[00m _[00m _[00m _[00m 
[01;40;31m*[00m _[00m [01;44;37mD[00m _[00m _[00m _[00m 
[01;40;31m*[00m _[00m _[00m _[00m _[00m [00;43;30m*[00m 
[01;40;31m*[00m _[00m _[00m [01;44;37mD[00m _[00m _[00m 
 
_[00m _[00m _[00m _[00m [01;44;37mD[00m _[00m 
[00;43;30m*[00m _[00m _[00m _[00m _[00m _[00m 
_[00m [00;43;30m*[00m _[00m _[00m _[00m _[00m 
[01;40;31mD[00m _[00m [01;44;37mD[00m _[00m _[00m _[00m 
[01;40;31mD[00m _[00m _[00m _[00m _[00m [00;43;30m*[00m 
[01;40;31mD[00m _[00m _[00m [01;44;37mD[00m _[00m _[00m 
 
_[00m _[00m _[00m _[00m [01;44;37mD[00m _[00m 
[00;43;30m*[00m _[00m _[00m _[00m _[00m _[00m 
_[00m [00;43;30m*[00m _[00m _[00m _[00m _[00m 
_[00m _[00m [01;44;37mD[00m _[00m _[00m _[00m 
_[00m _[00m _[00m _[00m _[00m [00;43;30m*[00m 
_[00m _[00m _[00m [01;44;37mD[00m _[00m _[00m 
 
